### ISSUE TITLE
Derived datasets: Bug in migration of "present" user_values

### ISSUE THREAD

**Issue by ploh** [quintel/etengine#904]
When migrating an existing scaled scenario (ID 530759) to a new derived dataset + new scenario based on the dataset, I tried moving all the "present" user_values into new initialiser inputs.
This should work since for each one there is a corresponding "future" user_value which should overwrite the change. But for some reason I get a lot of differing gquery results with this.
It works if I restrict myself to "present" user_values for which the "future" user_value has the same value, i.e. no sliders have been moved.

**Comment by ploh** [comment]
Including @grdw @ChaelKruip @antw

**Comment by ploh** [comment]
I am trying to find out which user_values exactly are causing the problem, now.

**Comment by ChaelKruip** [comment]
@AlexanderWirtz is this expected behaviour? 

@ploh could it be that the offending inputs are the those which gave @grdw a headache last week: https://github.com/quintel/etmodel/issues/2324

**Comment by AlexanderWirtz** [comment]
> I tried moving all the "present" user_values into new initialiser inputs.

@ploh: maybe a stupid question, but how about all the `_both` user_values? Shouldn't these be initialiser inputs as well?

**Comment by AlexanderWirtz** [comment]
@ChaelKruip I probably need to know a bit more before I can answer that. 

>But for some reason I get a lot of differing gquery results with this

I am curious what this means exactly

**Comment by ploh** [comment]
> @ploh could it be that the offending inputs are the those which gave @grdw a headache last week: quintel/etmodel#2324

Good idea. But if it is the same inputs, then only coincidentally. The `share_group` is not used for initializer inputs at all. And for scenarios' inputs it is only evaluated when the scenario is updated via the `ScenarioUpdater`, but I do not do that in my tests.

**Comment by ploh** [comment]
> > But for some reason I get a lot of differing gquery results with this

> I am curious what this means exactly

It means that I run _all_ gqueries from etsource and compare the results for the old and the new scenario. And there I get differences if I convert too many `present` user_values into initialiser inputs.

> > I tried moving all the "present" user_values into new initialiser inputs.

> @ploh: maybe a stupid question, but how about all the `_both` user_values? Shouldn't these be initialiser inputs as well?  

You are absolutely right, of course - but I am already doing that as well.

**Comment by ploh** [comment]
A lot of the errors were due to a stupid mistake I made which also deleted the "future" user_values.

But after fixing that, a few _present_ user_values are still causing problems. One of these is [`number_of_agriculture_chp_engine_network_gas_present`](https://github.com/quintel/etsource/blob/ea7cc5af9be787f76df760b01f32744e77b8f561/inputs/adjust_scaling/agriculture/number_of_agriculture_chp_engine_network_gas_present.ad). In Scenario 530759 it is set to 95.6387338082011. This is accompanied by setting [`number_of_agriculture_chp_engine_network_gas`](https://github.com/quintel/etsource/blob/ea7cc5af9be787f76df760b01f32744e77b8f561/inputs/demand/agriculture/number_of_agriculture_chp_engine_network_gas.ad) (the _future_ equivalent) to 0.0.

To make sure that the problems were not caused by some bug in the new derived dataset machinery, I simply created a _both_ input `number_of_agriculture_chp_engine_network_gas_both` (identical except for priority and update_period) and replaced the _present_ user_value by this.
Now, for the future graph first the _both_ input is executed and then the _future_ input - instead of only the _future_ input as previously.

I would have expected the later input to simply overwrite the former, but instead strange things happen:
![image](https://cloud.githubusercontent.com/assets/768025/22287274/6c1b0d3a-e2f2-11e6-8523-dc88c5501a2c.png)

Output of 2.48 B with an input of 0 does not seem to be physically possible :stuck_out_tongue_winking_eye: 

For a comparison the converter in the unchanged scenario:
![image](https://cloud.githubusercontent.com/assets/768025/22287362/e5004b48-e2f2-11e6-853e-ff290601f3f5.png)

And the converter where the output is flowing to - in the changed scenario:
![image](https://cloud.githubusercontent.com/assets/768025/22287407/0fdcfe4c-e2f3-11e6-8218-95fbfafd934a.png)

... and in the original scenario:
![image](https://cloud.githubusercontent.com/assets/768025/22287432/31312280-e2f3-11e6-929d-4d370aca5e3b.png)

@antw I asked @ChaelKruip about this and he did not really have an idea except that maybe the `production_based_on_number_of_units` in the update query might be using an old value!? Do you have any idea what might be causing this? Or how I can debug the execution of the queries?

**Comment by antw** [comment]
> maybe the production_based_on_number_of_units in the update query might be using an old value!?

That's possible. [Remove the `fetch` block in that method](https://github.com/quintel/etengine/blob/470256764853fa338a13543e272de324bd46610b/app/models/qernel/converter_api/capacity_production.rb#L69-L73) to prevent caching of the calculation, and see if it affects the outcome.

**Comment by antw** [comment]
> That's possible. Remove the fetch block in that method to prevent caching of the calculation, and see if it affects the outcome.

It occurs to me that, even if this is not the case, there *is* a problem here. If an initializer input uses a value which is cached by `fetch`, that cached value will persist even if a *normal* input would later cause it to change.

For example:

```ruby
def my_calculation
  fetch(:my_calculation)
    number_of_units + production_per_unit
  end
end
```

If I call `my_calculation` while executing an initializer input, the value will be remembered by `fetch`. If I later change the number of units using a *user* input, `my_calculation` will still have the old value. This is also true if you execute a `both` input followed by a `future` input; at least with init. inputs it should be possible for us to solve this.

This means that [calls to fetch](https://github.com/quintel/etengine/blob/master/app/models/qernel/dataset_attributes.rb#L162-L211) should not store any values during the initializer input phase and the block provided should always be executed. This would prevent inadvertently caching something too soon.

**Comment by ploh** [comment]
> > maybe the production_based_on_number_of_units in the update query might be using an old value!?

> That's possible. [Remove the `fetch` block in that method](https://github.com/quintel/etengine/blob/470256764853fa338a13543e272de324bd46610b/app/models/qernel/converter_api/capacity_production.rb#L69-L73) to prevent caching of the calculation, and see if it affects the outcome.

That worked :+1:

**Comment by ploh** [comment]
> This means that calls to fetch should not store any values during the initializer input phase and the block provided should always be executed. This would prevent inadvertently caching something too soon.

:+1: 
I will implement this. I wonder, however, why this problem never occurred before with normal user_values not correctly applying on top of each other!? Or maybe it did occur but no one noticed?

**Comment by ploh** [comment]
With the fetch caching disabled during the init. inputs application, most of the inputs do not cause a problem anymore, but... there are still some that do:
```
"transport_car_using_compressed_natural_gas_share_present"=>0.0559796358165971,
 "transport_car_using_diesel_mix_share_present"=>24.5580382529832,
 "transport_car_using_electricity_share_present"=>0.0297164625195247,
 "transport_car_using_gasoline_mix_share_present"=>71.0920435959933,
 "transport_car_using_hydrogen_share_present"=>0.01,
 "transport_car_using_lpg_share_present"=>4.25422205268735,
 "transport_car_using_compressed_natural_gas_share"=>0.0,
 "transport_car_using_diesel_mix_share"=>0.0,
 "transport_car_using_electricity_share"=>85.0,
 "transport_car_using_gasoline_mix_share"=>0.0,
 "transport_car_using_lpg_share"=>0.0,
 "transport_car_using_hydrogen_share"=>15.0
```

These user_values still cause a problem, i.e. when I convert all the `_present` user_values into init. inputs, I get some differing gquery results (though only 50 this time, not hundreds as before).

When I convert all but one to init. inputs, the differences disappear. I wonder what inside the input, e.g [for hydrogen share](https://github.com/quintel/etsource/blob/ea7cc5af9be787f76df760b01f32744e77b8f561/inputs/adjust_scaling/transport/transport_car_using_hydrogen_share_present.ad), might be causing this!?

@antw Do you have any idea?

**Comment by ploh** [comment]
Here are the [gquery result differences](https://gist.github.com/ploh/32e4db765f9c4177208eeb792c93bd8a)

**Comment by grdw** [comment]
Do queries on `number_of_unit` influence the `share` value? This might seem like a dumb question but as far as my guesswork goes: [here](https://github.com/quintel/etsource/blob/ea7cc5af9be787f76df760b01f32744e77b8f561/inputs/adjust_scaling/transport/transport_car_using_hydrogen_share_present.ad#L7-L12) is where a couple of relatable `transport_car_*` things are updated. Maybe some queries are now doing updating more than they should? This is just a guess though.

**Comment by ploh** [comment]
I figured out where the problem lies. It is in [this line of the inputs](https://github.com/quintel/etsource/blob/ea7cc5af9be787f76df760b01f32744e77b8f561/inputs/adjust_scaling/transport/transport_car_using_hydrogen_share_present.ad#L10) (they ones from this share group are all similar).
I believe there are two mistakes here:
1. The use of `INPUT_VALUE(transport_car_using_electricity_share)` although we are updating the present graph, i.e. it should be `INPUT_VALUE(transport_car_using_electricity_share_present)` here.
1. It is the same statement for the normal slider-connected future input. And there `QUERY_PRESENT(-> { AREA(number_of_cars) })` is a mistake since the future number of electric cars should be relative to the future total number of cars and not to the present total number.

The latter is also where the difference in the new derived dataset based scenario comes from: The `number_of_cars` is updated in this scenario for the present graph. I did not translate that into an init. input because it does not have a corresponding future input.
And when the above electric car statement is executed in the init. input, the total number is still the one from the ad file - instead of the updated one which will be updated only later when the normal user_values are applied.

After moving the `number_of_cars_present` user_value to the init. inputs, the gquery results are finally identical.

Open questions are how to deal with the three other present-only user_values from the original scenario. Leaving them in the user_values instead of init. inputs works but is kind of incorrect since we do not want to have _any_ present user_values in derived dataset based scenarios.
Moving them to init. inputs does not work for `transport_truck_using_lng_mix_share_present`.

The full set of user_values related to this:
```
@scenario.user_values.select {|k,_| k =~ /^transport_truck_using_/}
=> {"transport_truck_using_compressed_natural_gas_share_present"=>0.0663663397303467,
 "transport_truck_using_diesel_mix_share_present"=>99.078542447453,
 "transport_truck_using_electricity_share_present"=>0.00725330815142612,
 "transport_truck_using_gasoline_mix_share_present"=>0.847837904665262,
 "transport_truck_using_lng_mix_share_present"=>0,
 "transport_truck_using_compressed_natural_gas_efficiency"=>2.0,
 "transport_truck_using_compressed_natural_gas_share"=>0.0,
 "transport_truck_using_diesel_mix_share"=>15.71333333,
 "transport_truck_using_electricity_share"=>84.28666667,
 "transport_truck_using_gasoline_mix_share"=>0.0}
```

Sorry for maybe being not very clear and/or elaborate - it is late and I at least wanted to save my thoughts and investigations for myself here ;-)

**Comment by ploh** [comment]
> Moving them to init. inputs does not work for transport_truck_using_lng_mix_share_present.

I think this is probably a bug in the original scaled scenario (ID 530759): There should probably be a user_value
```
"transport_truck_using_lng_mix_share"=>0.0
```
as for the other carriers. In practice the omission of this leads to 
![image](https://cloud.githubusercontent.com/assets/768025/22324505/0a60f70e-e3aa-11e6-991c-4f714412e01f.png)
for the future input shares of [`transport_useful_demand_truck_kms`](https://beta-engine.energytransitionmodel.com/data/530759/converters/transport_useful_demand_truck_kms), i.e. `lng_mix` gets 0.09% instead of 0.0% (not much, but still causing some different gquery results down the road :wink:)

**Comment by ploh** [comment]
> I think this is probably a bug in the original scaled scenario (ID 530759)

@AlexanderWirtz Would appreciate your thoughts on this. And on the [above comment](https://github.com/quintel/etengine/issues/904#issuecomment-275199374).

**Comment by ploh** [comment]
Discussions earlier today with @AlexanderWirtz:
- The above 0.09% is probably a mistake. There should have been a future input as well.
- The above (about the `transport_car_...` shares) really seem to be errors in the inputs' queries, at least on first thought.

**Comment by ploh** [comment]
Mistakes in inputs are being solved in quintel/etsource#1223

Mistake in old Gea scenario 530759 is too minor to be solved.

Remember the encountered problems with `number_of_cars` and other _present-only_ inputs when migrating Paddepoel (quintel/etmodel#2319).

Close this now  :wink:
